ros2_batch_job/__main__.py

```diff
@@ -356,6 +366,89 @@ def build_and_test(args, job):
     # return 0 if ret_test == 0 and ret_testr == 0 else 1
     return 0
 
+'''
+Run colcon build on requested packages with appropriate cmake sanitizer flags
+'''
+def build_with_sanitizer(build_cmd, args, job):
+    print('# BEGIN SUBSECTION: build with sanitizer')
+
+    # Pass appropriate sanitizer flags
+    sanitizer_flag = '-fsanitize={}'.format(args.enable_sanitizer_type)
+    additional_compiler_flags = '-g'
+    sanitizer_cmake_args = [
+    '-DCMAKE_CXX_FLAGS="{} {}"'.format(sanitizer_flag, additional_compiler_flags),
+    '-DCMAKE_C_FLAGS="{} {}"'.format(sanitizer_flag, additional_compiler_flags),
+    '-DCMAKE_EXE_LINKER_FLAGS="{}"'.format(sanitizer_flag),
+    '-DCMAKE_MODULE_LINKER_FLAGS="{}"'.format(sanitizer_flag),
+    '-DCMAKE_SHARED_LINKER_FLAGS="{}"'.format(sanitizer_flag)
+    ]
+    add_cmake_args_to_cmd(build_cmd, sanitizer_cmake_args)
+
+    # Restrict build packages to the provided regex
+    if args.restrict_san_pkgs_regex:
+        build_cmd.extend(['--packages-select-regex ' + args.restrict_san_pkgs_regex])
+
+    ret_build = job.run(build_cmd, shell=True)
+    info("colcon build with sanitizer returned: '{0}'".format(ret_build))
+    print('# END SUBSECTION')
+    return ret_build
+
+'''
+Updates colcon test targets to only the required packages and adds pre-run/post-run
+commands for sanitizers
+'''
+def update_test_cmd_based_on_sanitizer_args(test_cmd, args):
+    test_cmd_prefix = get_test_cmd_prefix_based_on_sanitizer_args(args)
+    test_cmd_suffix = get_test_cmd_suffix_based_on_sanitizer_args(args)
+    return test_cmd_prefix + test_cmd + test_cmd_suffix
+
+'''
+Set up prerequisites for sanitizer tests and restrict tests to the required cmake packages
+'''
+def get_test_cmd_prefix_based_on_sanitizer_args(args):
+    '''
+    OSRF memory testing relies on LD_PRELOAD env variable to instrument libc functions like malloc().
+    Sanitizers work in a similar way and hence fail on memory tools tests with the message:
+    "ASan runtime does not come first in initial library list; you should either link runtime to your
+    application or manually preload it with LD_PRELOAD"
+
+    We are overriding LD_PRELOAD with the sanitizer library path before colcon test to resolve this.
+    FIXME(osrf/osrf_testing_tools_cpp#23): Add a cmake argument for disabling memory tool tests and
+    avoid overriding LD_PRELOAD.
+    '''
+    export_library_preload_for_sanitizer = []
+    if args.enable_sanitizer_type == 'address':
+        export_library_preload_for_sanitizer = ['export LD_PRELOAD=$(ls /usr/lib/x86_64-linux-gnu/libasan.so* | head -1);']
+
+    # Identify the sanitized packages
+    pkgs_regex_arg = '--packages-select-regex {}'.format(args.restrict_san_pkgs_regex) if args.restrict_san_pkgs_regex else ''
+    sanitized_pkgs = '$({} list {}'.format(args.colcon_script, pkgs_regex_arg) + '| awk \'{printf $1" "}\')'
```

Comment:
It would be more maintainable if this was done using list comprehension in python instead of going through a bash pipe to awk.